docs/state_encryption.md

```diff
@@ -0,0 +1,333 @@
+
+```hcl2
+terraform {
+  encryption {
+    key_provider "static" "foo" {
+      key = "6f6f706830656f67686f6834616872756f3751756165686565796f6f72653169"
+    }
+  }
+}
+```
+
+**Note:** the user is responsible for keeping this key safe and follow disaster recovery best practices. OpenTofu is a read-only consumer for the key the user provided and will not perform tasks on the key management system like key rotation.
+
+The user also has to specify at least one encryption method referencing the key provider. This encryption method determines how the encryption takes place. It is the user's responsibility to make sure that they provided a key that is suitable for the encryption method.
+
+```hcl2
+terraform {
+  encryption {
+    //...
+    method "aes_gcm" "bar" {
+      key_provider = key_provider.static.foo
+    }
+  }
+}
+```
+
+Finally, the user must reference the method for use in their state file, plan file, etc. This enables creating a different configuration for different purposes.
+
+```hcl2
+terraform {
+  encryption {
+    //...
+    statefile {
+      method = method.aes_gcm.abc
+    }
+    planfile {
+      method = method.aes_gcm.cde
+    }
+    backend {
+      method = method.pbkdf2.efg
+    }
+    remote_data_sources {
+      default {
+        method = method.aes_gcm.ghi
+      }
+      remote_data_source "some_module.remote_data_source.foo" {
+        method = method.aes_gcm.ijk
+      }
+    }
+  }
+}
+```
+
+To facilitate key and method rollover, the user can specify a fallback configuration for state and plan decryption. When the user specifies a `fallback` block, `tofu` will first attempt to decrypt any state or plan it reads with the primary method and then fall back to the `fallback` method. When `tofu` writes a state or plan, it will not use the `fallback` method and always writes with the primary method.
+
+```hcl2
+terraform {
+  encryption {
+    //...
+    statefile {
+      method = method.aes_gcm.bar
+      fallback_method {
+        method = method.aes_gcm.baz
+      } 
+    }
+  }
+}
+```
+
+**Note:** The `fallback` is a block because the future goals may require adding additional options.
+
+**Note:** Multiple `fallback` blocks should not be supported because they would be detrimental to performance and encourage keeping old encryption keys in the configuration.
+
+### Composite keys
+
+When a user desires to create a composite key, such as for creating a passphrase-based derivative key or a shared custody key, they may avail themselves of a key provider that supports multiple inputs. The user can chain key providers together:
+
+```hcl2
+terraform {
+  encryption {
+    key_provider "static" "my_passphrase" {
+      key = "this is my encryption key"
+    }
+    key_provider "pbkdf2" "my_derivative_key" {
+      key_providers = [key_provider.static.my_passphrase]
+    }
+    method "aes_gcm" "foo" {
+      key_provider = key_provider.pbkdf2.my_derivative_key
+    }
+    //...
+  }
+}
```

Comment:
I believe we've discussed this towards the beginning of this project - I'm not sure if exposing raw encryption details like "aes_gcm" or "pbkdf2" is great UX. I'm worried it could be confusing to users.

TBF, I think it would be better to have a simple "symmetric" method. In practice, I don't expect us to ever have any other methods, frankly. The underlying implementation details might of course change (e.g. from gcm to something else) but ideally we'd handle that transparently to the user (via fallback automatically inserted fallback methods).

In practice, I suppose we could use just "symmetric" for now, and in the future, if we need to, include separate parameters for the "aes" and "gcm" (mode_of_operation) and "256" (key_length) parts.

This is all based on my experience with the Go standard library, which exposes multiple modes of operation, and if you're not at least moderately well-versed with cryptography, you'll have a hard time deciding which one you should be using.

Reply:
Actually, when we were originally designing this, I believe we called the method "full" which I believe made sense.

It's obvious that we're doing symmetric encryption, and the fact that we're using aes_gcm should imo not be explicit in the config (or as an optional parameter).

And in the future, we might add a "partial" method.

Reply:
So, overall, the UX I would recommend, is the following:
```
terraform {
  encryption {
    key_provider "static" "key1" {
      hex = "6f6f706830656f67686f6834616872756f3751756165686565796f6f72653169"
    }
    key_provider "azure" "key2" {
      // Yada yada
    }
    method "full" "one" {
       key_provider = key_provider.static.key1
       mode = "gcm" // Optional, gcm is default.
       key_length = 256 // Optional.
    }
    method "partial" "two" {
       key_provider = key_provider.azure.key2
    }
    statefile {
      method = method.full.one
    }
    remote_data_sources {
      default {
        method = method.partial.two
      }
    }
  }
}
```

Reply:
Team choice made:

We will discuss how we can implement a 'simple' method in the future and what the UX would look like for that, as apposed to the current setup.